### PR TITLE
feat: Update random seed

### PR DESCRIPTION
**Pull request by evgenykuzyakov**
Fixes: https://github.com/near/near-sdk-rs/issues/223#issuecomment-688330778

Introduces a new protocol version that hashes the random seed per action using `action_hash` and `random_seed` from the block. Also updates the way receipt IDs are computed to avoid reusing IDs across forks.

# Test plan:
- [x] CI
- [x] nayduck http://nayduck.eastus.cloudapp.azure.com:3000/#/run/340
- new unittest

### CODE REVIEW

File: runtime/runtime/src/lib.rs [hunk 1]
```diff
@@ -497,6 +503,12 @@ impl Runtime {
         result.gas_burnt = exec_fee;
         // Executing actions one by one
         for (action_index, action) in action_receipt.actions.iter().enumerate() {
+            let action_hash = create_hash_upgradable(
+                apply_state.current_protocol_version,
+                &receipt.receipt_id,
+                &apply_state.last_block_hash,
+                u64::max_value() - action_index as u64,
```
Comment (bowenwang1996) [inline runtime/runtime/src/lib.rs:510]: why do we use `u64::max_value() - action_index` instead of `action_index`?

**Review by bowenwang1996 — COMMENTED** [review]
So action_hash will also change after the upgrade. Did we use it for anything?

File: runtime/runtime/src/lib.rs (hunk 1 above)
Comment (evgenykuzyakov) [reply runtime/runtime/src/lib.rs:510]: Otherwise it would match generated receipt IDs

Comment (evgenykuzyakov) [reply runtime/runtime/src/lib.rs:510]: I think originally we were using it for random values. But now we only use to generate unique `data_id` which is used to match data dependencies

Comment (nearmax) [inline runtime/runtime/src/lib.rs:510]: This implicitly encodes the type by counting action index backwards and receipt index and data index forwards. I guess the intention here is to make sure action hash does not overlap with generated receipt id. Unfortunately the calls to `create_hash_upgradable` are scattered across the code base so to make sure this does not happen one needs to find all usages of `create_hash_upgradable` and examine the way they are called, to make sure they do not produce overlapping ids.

I suggest we have 3 different implementations of `create_hash_upgradable` all located in one place and called something like `create_action_hash`, `create_data_id`, `create_receipt_id_from_receipt`, `create_receipt_id_from_transaction` to make it easy to examine visually that they produce non-overlapping ids. These implementation can also take strongly-typed arguments, e.g. `fn create_receipt_id_from_transaction(protocol_version: ProtocolVersion, transaction: &SignedTransaction, last_block_hash: &CryptoHash)` to hide stuff like `u64::max_value() - action_index as u64` internally and make sure it is not misused.

**Review by nearmax — CHANGES_REQUESTED** [review]

File: runtime/runtime/src/lib.rs (hunk 1 above)
Comment (evgenykuzyakov) [reply runtime/runtime/src/lib.rs:510]: I agree on having a separation for `create_receipt_id` vs `create_random_seed`, etc.

But we still need to use salt/nonce to generate receipt IDs for new receipts.

Comment (nearmax) [reply runtime/runtime/src/lib.rs:510]: Ideally, different types of salt like ActionIndex, ReceiptIndex, etc need to be strongly typed (`struct ActionIndex(u64)`), to not be misused, and compiler would optimize it out anyway. However, let's not do it in this PR but prepare it for this change by separating `create_receipt_id`, etc. Me and @frol were discussing the next steps towards the runtime spec, and the major step can be introducing strong typeness into the runtime, to make the methods as self-descriptive as possible.

Comment (evgenykuzyakov) [reply runtime/runtime/src/lib.rs]: Updated code to be explicit for every usage.

**Review by nearmax — APPROVED** [review]

File: core/primitives/src/utils.rs [hunk 2]
```diff
@@ -43,7 +49,105 @@ pub fn get_block_shard_id_rev(
...
+    protocol_version: ProtocolVersion,
+    receipt: &Receipt,
+    block_hash: &CryptoHash,
+    action_index: usize,
+) -> CryptoHash {
+    create_hash_upgradable(
+        protocol_version,
+        &receipt.receipt_id,
+        &block_hash,
+        u64::max_value() - action_index as u64,
```
Comment (nearmax) [inline core/primitives/src/utils.rs:85]: Please add a comment explaining what it does here.

File: core/primitives/src/utils.rs [hunk 3]
```diff
@@ -43,7 +49,105 @@ pub fn get_block_shard_id_rev(
     Ok((block_hash, shard_id))
 }
 
-pub fn create_nonce_with_nonce(base: &CryptoHash, salt: u64) -> CryptoHash {
+/// Creates a new Receipt ID from a given signed transaction and a block hash.
+/// This method is backward compatible, so it takes a current protocol version.
```
Comment (bowenwang1996) [inline core/primitives/src/utils.rs:53]:
```suggestion
/// This method is backward compatible, so it takes the current protocol version.
```

File: core/primitives/src/utils.rs [hunk 4]
```diff
@@ -43,7 +49,105 @@ pub fn get_block_shard_id_rev(
...
+pub fn create_receipt_id_from_transaction(
+    protocol_version: ProtocolVersion,
+    signed_transaction: &SignedTransaction,
+    block_hash: &CryptoHash,
+) -> CryptoHash {
+    create_hash_upgradable(protocol_version, &signed_transaction.get_hash(), &block_hash, 0)
+}
+
+/// Creates a new Receipt ID from a given receipt, a block hash and a new receipt index.
+/// This method is backward compatible, so it takes a current protocol version.
```
Comment (bowenwang1996) [inline core/primitives/src/utils.rs:63]:
```suggestion
/// This method is backward compatible, so it takes the current protocol version.
```

File: core/primitives/src/utils.rs [hunk 5]
```diff
@@ -43,7 +49,105 @@ pub fn get_block_shard_id_rev(
...
+    protocol_version: ProtocolVersion,
+    receipt: &Receipt,
+    block_hash: &CryptoHash,
+    receipt_index: usize,
+) -> CryptoHash {
+    create_hash_upgradable(protocol_version, &receipt.receipt_id, &block_hash, receipt_index as u64)
+}
+
+/// Creates a new action_hash from a given receipt, a block hash and an action index.
+/// This method is backward compatible, so it takes a current protocol version.
```
Comment (bowenwang1996) [inline core/primitives/src/utils.rs:74]:
```suggestion
/// This method is backward compatible, so it takes the current protocol version.
```

File: core/primitives/src/utils.rs [hunk 6]
```diff
@@ -43,7 +49,105 @@ pub fn get_block_shard_id_rev(
...
+    create_hash_upgradable(
+        protocol_version,
+        &receipt.receipt_id,
+        &block_hash,
+        u64::max_value() - action_index as u64,
+    )
+}
+
+/// Creates a new `data_id` from a given action hash, a block hash and a data index.
+/// This method is backward compatible, so it takes a current protocol version.
```
Comment (bowenwang1996) [inline core/primitives/src/utils.rs:90]:
```suggestion
/// This method is backward compatible, so it takes the current protocol version.
```

File: core/primitives/src/utils.rs [hunk 7]
```diff
@@ -43,7 +49,105 @@ pub fn get_block_shard_id_rev(
...
+    action_hash: &CryptoHash,
+    block_hash: &CryptoHash,
+    data_index: usize,
+) -> CryptoHash {
+    create_hash_upgradable(protocol_version, &action_hash, &block_hash, data_index as u64)
+}
+
+/// Creates a unique random seed to be provided to `VMContext` from a give `action_hash` and
+/// a given `random_seed`.
+/// This method is backward compatible, so it takes a current protocol version.
```
Comment (bowenwang1996) [inline core/primitives/src/utils.rs:102]:
```suggestion
/// This method is backward compatible, so it takes the current protocol version.
```

File: core/primitives/src/utils.rs [hunk 8]
```diff
@@ -43,7 +49,105 @@ pub fn get_block_shard_id_rev(
...
+pub fn create_random_seed(
+    protocol_version: ProtocolVersion,
+    action_hash: CryptoHash,
```
Comment (bowenwang1996) [inline core/primitives/src/utils.rs:105]: should we pass it as reference here?

**Review by bowenwang1996 — APPROVED** [review]

File: core/primitives/src/utils.rs (hunk 8 above)
Comment (evgenykuzyakov) [reply core/primitives/src/utils.rs:105]: `clippy` suggest to not pass structs by reference if they implement `Copy`. It's fixed 32 bytes, so it's okay to pass without a reference

File: core/primitives/src/utils.rs (hunk 2 above)
Comment (evgenykuzyakov) [reply core/primitives/src/utils.rs:85]: Added a comment

**Comment by abacabadabacaba** [comment]
Why is it important to have receipt ID unique across forks, while transaction hash doesn't have to be unique even within the same chain?

**Comment by evgenykuzyakov** [comment]
> Why is it important to have receipt ID unique across forks, while transaction hash doesn't have to be unique even within the same chain?

Receipt IDs are not as important, but Data IDs (for cross-contract dependencies) are also deducted from the ReceiptID.  Let's keep this discussion in the original reported issue.

**Comment by evgenykuzyakov** [comment]
Re-running nayduck: http://nayduck.eastus.cloudapp.azure.com:3000/#/run/351

**Pull request merged**